File: 1579.py
class Solution:
    def maxNumEdgesToRemove(self, n: int, edges: List[List[int]]) -> int:
        edgeTypes = [list() for _ in range(4)]
        for i, e in enumerate(edges):
            edgeTypes[e[0]].append(i)
        aParent = list(range(n + 1))
        bParent = list(range(n + 1))
        aCompSize = [1] * (n + 1)
        bCompSize = [1] * (n + 1)
        def getParent(x, pArr):
            while pArr[x] != x:
                x, pArr[x] = pArr[x], pArr[pArr[x]]
            return x
        def sameGroup(x, y, pArr):
            return getParent(x, pArr) == getParent(y, pArr)
        def unionComp(a, b, pArr, sArr):
            a = getParent(a, pArr)
            b = getParent(b, pArr)
            if a == b:
                return
            if sArr[a] < sArr[b]:
                a, b = b, a
            pArr[b] = a
            sArr[a] += sArr[b]
        chosenEdges = 0
        for i in edgeTypes[3]:
            _, x, y = edges[i]
            if sameGroup(x, y, aParent) and sameGroup(x, y, bParent):
                continue
            chosenEdges += 1
            unionComp(x, y, aParent, aCompSize)
            unionComp(x, y, bParent, bCompSize)
        if aCompSize[getParent(1, aParent)] != n:
            for i in edgeTypes[1]:
                _, x, y = edges[i]
                if sameGroup(x, y, aParent):
                    continue
                chosenEdges += 1
                unionComp(x, y, aParent, aCompSize)
        if aCompSize[getParent(1, aParent)] != n:
            return -1
        if bCompSize[getParent(1, bParent)] != n:
            for i in edgeTypes[2]:
                _, x, y = edges[i]
                if sameGroup(x, y, bParent):
                    continue
                chosenEdges += 1
                unionComp(x, y, bParent, bCompSize)
        if bCompSize[getParent(1, bParent)] != n:
            return -1
        else:
            return len(edges) - chosenEdges

File: sol/1579.cpp
struct Solution {
    int getParent(int x, vector<int> &pArr){
        while (pArr[x] != x){
            int y = x;
            x = pArr[y];
            pArr[y] = pArr[pArr[y]];
        }
        return x;
    }
    inline bool sameGroup(int x, int y, vector<int> &pArr){
        return getParent(x, pArr) == getParent(y, pArr);
    }
    inline void unionComp(int x, int y, vector<int> &pArr, vector<int> &sArr){
        x = getParent(x, pArr);
        y = getParent(y, pArr);
        if (x == y){
            return;
        }
        if (sArr[x] < sArr[y]){
            swap(x, y);
        }
        pArr[y] = x;
        sArr[x] += sArr[y];
    }
    int maxNumEdgesToRemove(int n, vector<vector<int>>& edges) {
        vector<int> edgeGp[4];
        for (int i = 0; i < edges.size(); ++i){
            edgeGp[edges[i][0]].push_back(i);
        }
        vector<int> aPar(n + 1), bPar(n + 1), aSize(n + 1, 1), bSize(n + 1, 1);
        iota(aPar.begin(), aPar.end(), 0);
        iota(bPar.begin(), bPar.end(), 0);
        int count = 0, x, y;
        for (int &i : edgeGp[3]){
            x = edges[i][1];
            y = edges[i][2];
            if (sameGroup(x, y, aPar) && sameGroup(x, y, bPar)){
                continue;
            }
            ++count;
            unionComp(x, y, aPar, aSize);
            unionComp(x, y, bPar, bSize);
        }
        if (aSize[getParent(1, aPar)] != n){
            for (int &i : edgeGp[1]){
                x = edges[i][1];
                y = edges[i][2];
                if (sameGroup(x, y, aPar)){
                    continue;
                }
                ++count;
                unionComp(x, y, aPar, aSize);
            }
        }
        if (aSize[getParent(1, aPar)] != n){
            return -1;
        }
        if (bSize[getParent(1, bPar)] != n){
            for (int &i : edgeGp[2]){
                x = edges[i][1];
                y = edges[i][2];
                if (sameGroup(x, y, bPar)){
                    continue;
                }
                ++count;
                unionComp(x, y, bPar, bSize);
            }
        }
        if (bSize[getParent(1, bPar)] != n){
            return -1;
        } else {
            return edges.size() - count;
        }
    }
};
